bintray/services/upload.go

```diff
@@ -76,39 +79,69 @@ func (us *UploadService) uploadFiles(uploadDetails *UploadParams, artifacts []cl
 
 	// Create an array of integers, to store the total file that were uploaded successfully.
 	// Each array item is used by a single thread.
-	uploadCount := make([]int, us.Threads, us.Threads)
+	uploadedArtifacts := make(map[int][]clientutils.Artifact, us.Threads)
 	matrixParams := getMatrixParams(uploadDetails)
 	for i := 0; i < us.Threads; i++ {
 		wg.Add(1)
+		// The threadId key MUST exist in the map first so that the map itself is not being mutated inside a go routine
+		uploadedArtifacts[i] = make([]clientutils.Artifact, 0)
 		go func(threadId int) {
 			logMsgPrefix := clientutils.GetLogMsgPrefix(threadId, us.DryRun)
 			for j := threadId; j < size; j += us.Threads {
-				if err != nil {
-					break
-				}
-				url := baseUrl + "/" + artifacts[j].TargetPath + matrixParams
 				if !us.DryRun {
+					url := baseUrl + "/" + artifacts[j].TargetPath + matrixParams
 					uploaded, e := uploadFile(artifacts[j], url, logMsgPrefix, us.BintrayDetails)
 					if e != nil {
-						err = e
-						break
+						log.Error(logMsgPrefix, "Failed uploading artifact:", artifacts[j].LocalPath, ":", e)
 					}
 					if uploaded {
-						uploadCount[threadId]++
+						uploadedArtifacts[threadId] = append(uploadedArtifacts[threadId], artifacts[j])
 					}
 				} else {
 					log.Info("[Dry Run] Uploading artifact:", artifacts[j].LocalPath)
-					uploadCount[threadId]++
+					uploadedArtifacts[threadId] = append(uploadedArtifacts[threadId], artifacts[j])
 				}
 			}
 			wg.Done()
 		}(i)
 	}
 	wg.Wait()
 
+	if uploadDetails.ShowInDownloadList {
+		// even though we are not running the list for download in go routines we need this outer loop
```

Comment:
Great comment. For consistency reasons., let's start with an upper-case (even --> Even).